src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -983,100 +724,50 @@ private void EnsureSessionHandleExists(RequestState state)
                             accessType,
                             Interop.WinHttp.WINHTTP_NO_PROXY_NAME,
                             Interop.WinHttp.WINHTTP_NO_PROXY_BYPASS,
-                            0);
-                        if (!_sessionHandle.IsInvalid)
+                            (int)Interop.WinHttp.WINHTTP_FLAG_ASYNC);
+                        if (_sessionHandle.IsInvalid)
```

Comment:
This is a bit confusing. The code would be easier to read if it would first check for `ERROR_INVALID_PARAMETER` and, if true, do the WinINetProxy fallback block. 
Afterwards, it could call ThrowOnInvalidHandle() outside of any `if` blocks since the helper will internally check the handle parameter for validity.